File: docs-examples/example-java/src/test/java/io/micronaut/mqtt/docs/parameters/BindingSpec.java
package io.micronaut.mqtt.docs.parameters;

import io.micronaut.mqtt.AbstractMQTTTest;
import io.micronaut.context.ApplicationContext;
import org.junit.jupiter.api.Test;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

public class BindingSpec extends AbstractMQTTTest {

    @Test
    void testDynamicBinding() {
        ApplicationContext applicationContext = startContext();

// tag::producer[]
        ProductClient productClient = applicationContext.getBean(ProductClient.class);
        productClient.send("message body".getBytes());
        productClient.send("product/a", "message body2".getBytes());
        productClient.send("product/b", "message body2".getBytes());
// end::producer[]

        ProductListener productListener = applicationContext.getBean(ProductListener.class);

        try {
            await().atMost(5, SECONDS).until(() ->
                    productListener.messageLengths.size() == 1 &&
                    productListener.messageLengths.contains(12) &&
                    productListener.topics.size() == 2 &&
                    productListener.topics.get(0).equals("product/a") &&
                    productListener.topics.get(1).equals("product/b")
            );
        } finally {
            applicationContext.close();
        }
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/mqtt/docs/parameters/BindingSpec.kt
package io.micronaut.mqtt.docs.parameters

import io.micronaut.mqtt.AbstractMQTTTest
import org.awaitility.Awaitility.await
import org.junit.jupiter.api.Test
import java.util.concurrent.TimeUnit

class BindingSpec : AbstractMQTTTest() {

    @Test
    fun testProducerAndConsumer() {
        val applicationContext = startContext()

        // tag::producer[]
        val productClient = applicationContext.getBean(ProductClient::class.java)
        productClient.send("message body".toByteArray())
        productClient.send("product", "message body2".toByteArray())
        // end::producer[]

        val productListener = applicationContext.getBean(ProductListener::class.java)

        try {
            await().atMost(5, TimeUnit.SECONDS).until {
                productListener.messageLengths.size == 2 &&
                        productListener.messageLengths.contains(12) &&
                        productListener.messageLengths.contains(13)
            }
        } finally {
            applicationContext.close()
        }
    }
}
